File: Client/src/types/note.ts
import type { NoteCanvas, NoteDocType } from './noteCanvas';

export type NoteKind = 'session_prep' | 'narrative';

/** The protected folders every campaign gets. 'dm_notes' is a CHILD of the 'session' folder
 * and holds the DM's chat threads (SessionChat) rather than Notes - see NotesFolderExplorer,
 * which renders chat rows when it is the open folder. */
export type NoteFolderDefaultKind = 'session' | 'narrative' | 'dm_notes';

export interface Note {
  id: string;
  campaignId: string;
  folderId: string | null;
  name: string;
  kind: NoteKind | null;
  /** Which editor this file opens in: 'text' (body), 'whiteboard' or 'tree' (canvas).
   * Every note written before the canvas types existed reads back as 'text'. */
  docType: NoteDocType;
  /** HTML - the document itself for a 'text' note, unused by the canvas types. */
  body: string;
  /** The whiteboard/tree document, normalized on the way in (see noteCanvas.ts). null for a
   * text note. */
  canvas: NoteCanvas | null;
  tags: string[];
  createdAt: number;
  updatedAt: number;
}

export interface NoteFolder {
  id: string;
  campaignId: string;
  parentId: string | null;
  name: string;
  isDefault: boolean;
  defaultKind: NoteFolderDefaultKind | null;
  createdAt: number;
  updatedAt: number;
}

File: Server/app/schemas/note.py
import uuid
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator

NoteKind = Literal["session_prep", "narrative"]
# Which editor a note opens in - see Note.doc_type. Orthogonal to NoteKind.
NoteDocType = Literal["text", "whiteboard", "tree"]
# "dm_notes" is the protected folder INSIDE the session folder that lists the DM's chat
# threads (issues.txt 10.b.3) rather than notes - see the client's NotesFolderExplorer.
NoteFolderDefaultKind = Literal["session", "narrative", "dm_notes"]


class NoteFolderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    campaign_id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    name: str
    is_default: bool
    default_kind: Optional[NoteFolderDefaultKind] = None
    created_at: datetime
    updated_at: datetime


class NoteFolderCreate(BaseModel):
    id: Optional[uuid.UUID] = None
    campaign_id: uuid.UUID
    parent_id: Optional[uuid.UUID] = None
    name: str
    is_default: bool = False
    default_kind: Optional[NoteFolderDefaultKind] = None


class NoteFolderUpdate(BaseModel):
    parent_id: Optional[uuid.UUID] = None
    name: Optional[str] = None


class NoteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    campaign_id: uuid.UUID
    folder_id: Optional[uuid.UUID] = None
    name: str
    kind: Optional[NoteKind] = None
    doc_type: NoteDocType = "text"
    body: str
    canvas: Optional[Any] = None
    tags: Optional[Any] = None
    created_at: datetime
    updated_at: datetime


class _CanvasNeverNull(BaseModel):
    """notes.canvas is NOT NULL (server_default '{}'), matching notes.tags - but the client
    sends the whole note back on every PATCH, and a text note's canvas is null on its side.
    Coerce it here rather than making the column nullable, so "no canvas" has exactly one
    representation in the database."""

    @field_validator("canvas", check_fields=False)
    @classmethod
    def _empty_canvas(cls, value: Optional[Any]) -> Any:
        return {} if value is None else value


class NoteCreate(_CanvasNeverNull):
    id: Optional[uuid.UUID] = None
    campaign_id: uuid.UUID
    folder_id: Optional[uuid.UUID] = None
    name: str
    kind: Optional[NoteKind] = None
    doc_type: NoteDocType = "text"
    body: str = ""
    # default_factory, not None: a validator does not run on a default, and the column is
    # NOT NULL - a create that omits canvas has to arrive as {}.
    canvas: Any = Field(default_factory=dict)
    tags: Optional[Any] = None


class NoteUpdate(_CanvasNeverNull):
    folder_id: Optional[uuid.UUID] = None
    name: Optional[str] = None
    kind: Optional[NoteKind] = None
    doc_type: Optional[NoteDocType] = None
    body: Optional[str] = None
    canvas: Optional[Any] = None
    tags: Optional[Any] = None
